JDK-8287061: Support for rematerializing scalar replaced objects participating in allocation merges

JohnTortugo opened the pull request:
Can I please get reviews for this PR? 

The most common and frequent use of NonEscaping Phis merging object allocations is for debugging information. The two graphs below show numbers for Renaissance and DaCapo benchmarks - similar results are obtained for all other applications that I tested.

With what frequency does each IR node type occurs as an allocation merge user? I.e., if the same node type uses a Phi N times the counter is incremented by N:

![image](https://user-images.githubusercontent.com/2249648/222280517-4dcf5871-2564-4207-b49e-22aee47fa49d.png)

What are the most common users of allocation merges? I.e., if the same node type uses a Phi N times the counter is incremented by 1:

![image](https://user-images.githubusercontent.com/2249648/222280608-ca742a4e-1622-4e69-a778-e4db6805ea02.png)

This PR adds support scalar replacing allocations participating in merges used as debug information OR as a base for field loads. I plan to create subsequent PRs to enable scalar replacement of merges used by other node types (CmpP is next on the list) subsequently.

The approach I used for _rematerialization_ is pretty straightforward. It consists basically of the following. 1) New IR node (suggested by V. Kozlov), named SafePointScalarMergeNode, to represent a set of SafePointScalarObjectNode; 2) Each scalar replaceable input participating in a merge will get a SafePointScalarObjectNode like if it weren't part of a merge. 3) Add a new Class to support the rematerialization of SR objects that are part of a merge; 4) Patch HotSpot to be able to serialize and deserialize debug information related to allocation merges; 5) Patch C2 to generate unique types for SR objects participating in some allocation merges.

The approach I used for _enabling the scalar replacement of some of the inputs of the allocation merge_ is also pretty straightforward: call `MemNode::split_through_phi` to, well, split AddP->Load* through the merge which will render the Phi useless.

I tested this with JTREG tests tier 1-4 (Windows, Linux, and Mac) and didn't see regression. I also experimented with several applications and didn't see any failure. I also ran tests with "-ea -esa -Xbatch -Xcomp -XX:+UnlockExperimentalVMOptions -XX:-TieredCompilation -server -XX:+IgnoreUnrecognizedVMOptions -XX:+UnlockDiagnosticVMOptions -XX:+StressLCM -XX:+StressGCM -XX:+StressCCP" and didn't observe any related failures.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287061](https://bugs.openjdk.org/browse/JDK-8287061): Support for rematerializing scalar replaced objects participating in allocation merges (**Enhancement** - P4)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12897/head:pull/12897` \
`$ git checkout pull/12897`

Update a local copy of the PR: \
`$ git checkout pull/12897` \
`$ git pull https://git.openjdk.org/jdk.git pull/12897/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12897`

View PR using the GUI difftool: \
`$ git pr show -t 12897`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12897.diff">https://git.openjdk.org/jdk/pull/12897.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12897#issuecomment-1457364821)

JohnTortugo commented:
/label add hotspot-runtime

merykitty commented:
/label hotspot-compiler

vnkozlov reviewed (commented):
Thank you, @JohnTortugo, for continue working on it. I will test it and do proper review late.
Of cause @iwanowww have to approve it too :)

vnkozlov commented:
You new test failed in GHA testing with 32-bit VM: `Could not find VM flag "UseCompressedOops" in @IR rule 1 at int`.
You need to adjust next rule: `@IR(counts = { IRNode.ALLOC, "2" }, applyIf = { "UseCompressedOops", "false" })`

vnkozlov commented on src/hotspot/share/code/debugInfo.hpp:
```diff
@@ -165,6 +187,63 @@ class ObjectValue: public ScopeValue {
+// of a candidate object in '_possible_objects'. That candidate object is an
+// ObjectValue describing an object that was scalar replaced.
+
+class ObjectMergeValue: public ScopeValue {
```
Why you did not make subclass of ObjectValue? You would need to check `sv->is_object_merge()` first before `sv->is_object()` in few places. But on other hand you don't need to duplicates ObjectValue`s fields and asserts.

JohnTortugo replied:
Let me try that and see how it looks.

JohnTortugo replied:
Hi @vnkozlov, just FYI. I made the changes that you suggested. Please let me know what you think.

vnkozlov commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -1465,17 +1468,21 @@ SafePointScalarObjectNode::SafePointScalarObjectNode(const TypeOopPtr* tp,
-  _n_fields(n_fields)
+  _n_fields(n_fields),
 #ifdef ASSERT
-  , _alloc(alloc)
+  _alloc(alloc),
 #endif
```
May be we should always pass alloc, even in product VM. It is not related to your changes but it is pain to have.

JohnTortugo replied:
I can make that change.

vnkozlov commented on src/hotspot/share/opto/callnode.hpp:
```diff
@@ -503,15 +503,76 @@ class SafePointNode : public MultiNode {
+//   _first_index : relative index in the SafePoint node's input array where pointers
+//                  to the values of the fields of the SR object are going to be stored.
+//   _nfields     : how many fields the SR object has.
+//   _alloc       : pointer to the Allocate object that previously created the SR object.
+//                  Only used for debug purposes.
```
May be useful in other cases too in a future not only in debug.

vnkozlov commented on src/hotspot/share/opto/macro.hpp:
```diff
@@ -196,7 +193,7 @@ class PhaseMacroExpand : public Phase {
                           Node* size_in_bytes);
 
-  Node* make_arraycopy_load(ArrayCopyNode* ac, intptr_t offset, Node* ctl, Node* mem, BasicType ft, const Type *ftype, AllocateNode *alloc);
+  static Node* make_arraycopy_load(Compile* comp, PhaseIterGVN* igvn, ArrayCopyNode* ac, intptr_t offset, Node* ctl, Node* mem, BasicType ft, const Type *ftype, AllocateNode *alloc);
```
Why you need this change? It polluted diffs and hide important changes. Could be separate change from this one.

JohnTortugo replied:
I had to make this method static because it uses `value_from_mem` - which I also made static. I had to make `value_from_mem` static so that I can use it outside PhaseMacroExpand.

vnkozlov replied:
I see, you use in escape.cpp. Okay. I need to review changes there too.

navyxliu replied:
or you could construct a temporary PhaseMacroExpand object in EA. 

I see that you convert many member function to static so you can query in EA. the only blocker is _igvn.

JohnTortugo replied:
That seems a good idea. Together with some other refactoring I decided to revert making the methods static and instead use them through an instance of ME.

vnkozlov commented on src/hotspot/share/opto/callnode.hpp:
```diff
@@ -503,15 +503,76 @@ class SafePointNode : public MultiNode {
+// A SafePointScalarObjectNode (aka spobj) represents the state of one or more
+// scalarized objects at a SafePointNode. There are two scenarios where a
+// 'spobj' is created: 1) A scalar replaced (SR) object is directly referenced
+// by a SafePoint; 2) A scalar replaced object is participating in an allocation
+// merge (Phi) and the Phi is referenced by a SafePoint. The schematics of how
+// 'spobj' is used in both scenarios are described below.
```
I am not comfortable with reusing SafePointScalarObjectNode for 2) since it describes totally different information.
I think it should be separate Node which points to array of  SFSO  id (in addition to Phis) similar how we do now if SFSO is referenced in other SFSO's field.   SFSO could be created before the merge. Consider:
```
  Point p = new Point();
  Point q = foo();
  if (cond) {
      q = p;
  }
  trap(p, q);
```

JohnTortugo replied:
I had considered that but decided not to do it to prevent adding a new IR node. I'll give that a shot and update this thread with how it goes.

vnkozlov replied:
It **will** complicate your DebugInfo code (packing/unpacking) information. But I think it is right thing to do to avoid duplicated re-allocations during deoptimization.

JohnTortugo replied:
Hi @vnkozlov, just FYI. I made the changes that you suggested.

vnkozlov reviewed (commented):
Initial review.

navyxliu commented on src/hotspot/share/opto/callnode.hpp:
```diff
@@ -542,6 +607,20 @@ class SafePointScalarObjectNode: public TypeNode {
+
+  int merge_pointer_idx(JVMState* jvms) const {
+    assert(jvms != nullptr, "JVMS reference is null.");
+    return jvms->scloff() + _merge_pointer_idx;
```
how about we also assert is_from_merge() here? 

Your comment above says that _merge_point_idx is a zero-based index of sfpt's input array. 
here we use scloff-based. I think either is okay, but we need consistency.

JohnTortugo replied:
I think adding assert is a good idea. I'll fix the comment. Thanks!

JohnTortugo commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3350,6 +3728,12 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
+      // Reducible Phi's will be removed from the graph after split_unique_types finishes
+      if (reducible_merges.member(n)) {
+        // Split loads through phi
+        reduce_this_phi_on_field_access(n->as_Phi(), alloc_worklist);
```
I decided to do the split here so that Phase 1 of `split_unique_types` could assign a new instance type for the new loads created by `split_through_phi`. However, I'm considering doing the split at the end of `compute_escape` instead of here to keep all the code that does split close together.

navyxliu commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -401,6 +430,329 @@ bool ConnectionGraph::compute_escape() {
+  }
+}
+
+// This method will create a SafePointScalarObjectNode for each combination of
```
you have changed to SafePointScalarMergeNode in code.

navyxliu commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -626,9 +626,11 @@ bool PhaseMacroExpand::can_eliminate_allocation(AllocateNode *alloc, GrowableArr
+        } else if (safepoints != nullptr) {
+          safepoints->append_if_missing(sfpt);
         }
+      } else if (ignore_merges && (use->is_Phi() || use->is_EncodeP() || use->Opcode() == Op_MemBarRelease)) {
```
I try to understand this part. now `can_eliminate_allocation` can pre-test whether SR can eliminate the alloc.  I see that you use it in EA. 

With ignore_merges, why we also skip EncodeP or MemBarRelease here?

navyxliu replied:
Do you really need the boolean parameter ignore_merges here? 
It looks like we can use (safepoints == nullptr) instead?

navyxliu commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -401,6 +430,329 @@ bool ConnectionGraph::compute_escape() {
+    // Add the selector so we know which direction the execution took
+    call->add_req(selector);
+
+    for (uint i = 1; i < ophi->req(); i++) {
```
Comparing to new_phi and selector, I think this is the heavy-lifting work. You "replace" all appearances of ptn with SPSO. This logic almost overlaps the 'scalar replacement' part in MacroExpand. 

Do you consider to perform the transformation in MacroExpand? Your prior changes have already removed NSR marks, ME/SR will consider 'ptn'.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -401,6 +430,298 @@ bool ConnectionGraph::compute_escape() {
+    // Replaces debug information references to "ophi" in "sfpt" with references to "smerge"
+    int debug_end = jvms->debug_end();
+    sfpt->replace_edges_in_range(ophi, smerge, debug_start, debug_end, _igvn);
+    sfpt->set_req(smerge->merge_pointer_idx(jvms), ophi);
```
So you trying to restore `ophi` in debug info which was added at line 637 but then in previous line may be replaced with `smerge`. May add  comment explaining that.

vnkozlov commented on src/hotspot/share/opto/output.cpp:
```diff
@@ -764,6 +764,32 @@ void PhaseOutput::FillLocArray( int idx, MachSafePointNode* sfpt, Node *local,
+      set_sv_for_object_node(objs, sv);
+
+      for (uint i = 1; i < smerge->req(); i++) {
+        Node* fld_node = smerge->in(i);
```
It is not `fld_node` but `obj_node`.

vnkozlov commented on src/hotspot/share/opto/output.cpp:
```diff
@@ -748,14 +748,14 @@ void PhaseOutput::FillLocArray( int idx, MachSafePointNode* sfpt, Node *local,
       ciKlass* cik = t->is_oopptr()->exact_klass();
       assert(cik->is_instance_klass() ||
-             cik->is_array_klass(), "Not supported allocation.");
+            cik->is_array_klass(), "Not supported allocation.");
```
Why spacing changed?

JohnTortugo replied:
The identation level was incorrect before.

vnkozlov reviewed (commented):
Thank you for adding new node - it is more clear now.

JohnTortugo commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -401,6 +430,298 @@ bool ConnectionGraph::compute_escape() {
+    int debug_start     = jvms->debug_start();
+
+    SafePointScalarMergeNode* smerge = new SafePointScalarMergeNode(merge_t, merge_idx);
+    smerge->init_req(0, _compile->root());
```
To be honest, for this Node, I thought it didn't matter. I actually just used the same pattern as in PhaseMacroExpand.
I'll adjust the patch as you suggested, though.

JohnTortugo commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -401,6 +430,329 @@ bool ConnectionGraph::compute_escape() {
+      if (PhaseMacroExpand::can_eliminate_allocation(_igvn, alloc, nullptr, true)) {
+        found_sr_allocate = true;
+      } else {
+        ptn->set_scalar_replaceable(false);
```
This is actually necessary here. By setting the input to NSR I don't need to later, when performing reduction, check that I can eliminate the node. I can just check that I can scalar replace the input. If I removed this line I'd hit a problem if the merge had an input that is SR but that ME can't eliminate.

navyxliu replied:
okay. I see you mean.

JohnTortugo commented:
@vnkozlov - I think I addressed all your comments. Please let me know if I missed something or if there is something on that you think need to be improved.

@iwanowww - can I ask you to please take a look and let me know what you think?

vnkozlov reviewed (commented):
Nice. I will test it.

vnkozlov commented:
New test failed in tier1 on all platforms. Here is list:
```
1) Method "int compiler.c2.irTests.scalarReplacement.AllocationMergesTests.TestTrapAfterMerge(boolean,int,int)" - [Failed IR rules: 1]:
2) Method "int compiler.c2.irTests.scalarReplacement.AllocationMergesTests.testCondLoadAfterMerge(boolean,boolean,int,int)" - [Failed IR rules: 1]:
3) Method "int compiler.c2.irTests.scalarReplacement.AllocationMergesTests.testLoadInCondAfterMerge(boolean,int,int)" - [Failed IR rules: 1]:
4) Method "int compiler.c2.irTests.scalarReplacement.AllocationMergesTests.testMergesAndMixedEscape(boolean,int,int)" - [Failed IR rules: 1]:
5) Method "compiler.c2.irTests.scalarReplacement.AllocationMergesTests$Point[] compiler.c2.irTests.scalarReplacement.AllocationMergesTests.testNestedObjectsArray(boolean,int,int)" - [Failed IR rules: 1]:
6) Method "int compiler.c2.irTests.scalarReplacement.AllocationMergesTests.testNestedObjectsNoEscapeObject(boolean,int,int)" - [Failed IR rules: 1]:
7) Method "compiler.c2.irTests.scalarReplacement.AllocationMergesTests$Point compiler.c2.irTests.scalarReplacement.AllocationMergesTests.testNestedObjectsObject(boolean,int,int)" - [Failed IR rules: 1]:
```

JohnTortugo commented:
@vnkozlov - sorry about that. I fixed the code now and all GHA tests are passing.

vnkozlov commented:
Submitted new testing

vnkozlov commented:
Again got failures in the test on Aarch64 running with -XX:-UseTLAB:
```
testCmpMergeWithNull(boolean,int,int):
- Failed comparison: [found] 0 = 2 [given]
testCmpMergeWithNull_Second(boolean,int,int)
- Failed comparison: [found] 0 = 1 [given]
testMergedAccessAfterCallNoWrite(boolean,int,int)
- Failed comparison: [found] 2 = 3 [given]
testMergedAccessAfterCallWithWrite(boolean,int,int)
- Failed comparison: [found] 2 = 3 [given]
testNestedObjectsArray(boolean,int,int)
- Failed comparison: [found] 2 = 4 [given]
```

vnkozlov commented:
Also next 2 JVMCI tests failed:
```
compiler/jvmci/jdk.vm.ci.runtime.test/src/jdk/vm/ci/runtime/test/TestResolvedJavaMethod.java
compiler/jvmci/jdk.vm.ci.runtime.test/src/jdk/vm/ci/runtime/test/TestResolvedJavaType.java

#  Internal Error (/workspace/open/src/hotspot/cpu/x86/macroAssembler_x86.cpp:829), pid=2430194, tid=2430218
#  fatal error: DEBUG MESSAGE: exact klass and actual klass differ
```
Could be due to [12810](https://git.openjdk.org/jdk/pull/12810)

JohnTortugo commented:
Thank you for testing, Vladimir. I was able to reproduce the IR test failures on AArch64 with -UseTLAB. I'll push a fix later today. 

Looks like the other failures are due to: https://bugs.openjdk.org/browse/JDK-8306581

JohnTortugo commented:
> Again got failures in the test on Aarch64 running with -XX:-UseTLAB:
> 
> ```
> testCmpMergeWithNull(boolean,int,int):
> - Failed comparison: [found] 0 = 2 [given]
> testCmpMergeWithNull_Second(boolean,int,int)
> - Failed comparison: [found] 0 = 1 [given]
> testMergedAccessAfterCallNoWrite(boolean,int,int)
> - Failed comparison: [found] 2 = 3 [given]
> testMergedAccessAfterCallWithWrite(boolean,int,int)
> - Failed comparison: [found] 2 = 3 [given]
> testNestedObjectsArray(boolean,int,int)
> - Failed comparison: [found] 2 = 4 [given]
> ```

@vnkozlov - The reason for these failures is due to an issue in the test framework ALLOC Regex: https://bugs.openjdk.org/browse/JDK-8306625 . Since only the tests added in this PR are failing due to that problem do you think I should create a separate PR to fix the Regex or just include the fix in this PR?

vnkozlov commented:
> Since only the tests added in this PR are failing due to that problem do you think I should create a separate PR to fix the Regex or just include the fix in this PR?

Create separate PR and fix it first. This PR still need review from @iwanowww and it may take time to address additional comments.

iwanowww commented on src/hotspot/share/code/debugInfo.cpp:
```diff
@@ -191,6 +216,104 @@ void ObjectValue::print_fields_on(outputStream* st) const {
+
+  // If we call select again on the same merge we should return the same result
+  if (_selected != nullptr) {
+    return _selected;
```
I'm not sure I understand how it is intended to work. The code below initializes `_selected`, but returns `nullptr` when `selector >= 0`. Subsequent calls will return non-null value.

JohnTortugo replied:
This can be improved. I'll fix it.

iwanowww commented on src/hotspot/share/code/debugInfo.hpp:
```diff
@@ -165,13 +187,59 @@ class ObjectValue: public ScopeValue {
+// of a candidate object in '_possible_objects'. That candidate object is an
+// ObjectValue describing an object that was scalar replaced.
+
+class ObjectMergeValue: public ObjectValue {
```
I find the decision to subclass`ObjectValue` confusing and error prone: now `is_object()` returns true for `ObjectMergeValue`, but you have to apply the selector first to turn it into `ObjectValue`. And now the order of checks matter, so you always have to perform `is_object_merge()` first and then follow it with `is_object()` guard.

You have 3 flavors of `ObjectValue` now:
* good old `ObjectValue`;
* `ObjectMergeValue`
* merge candidates (`ObjectMergeCandidateValue`?)

Does it make sense to introduce 3 different subclasses under `ObjectValue` to clearly distinguish the scenarios?

JohnTortugo replied:
> Does it make sense to introduce 3 different subclasses under ObjectValue to clearly distinguish the scenarios?

I think that's a good idea. I'll give it a shot. Thanks.

iwanowww replied:
Can `ObjectCandidateValue` be a wrapper around a `ObjectAllocationValue`?

It does make sense to separate `ObjectMergeValue` and `ObjectValue`.

iwanowww replied:
I need to to study the code in more details. Seems like I'm missing something important here.

JohnTortugo replied:
@iwanowww - how can I make it easier for you to review? Thanks for your comments so far.

iwanowww commented on src/hotspot/share/opto/escape.hpp:
```diff
@@ -578,6 +587,17 @@ class ConnectionGraph: public ArenaObj {
+  // -------------------------------------------
+  // Methods related to Reduce Allocation Merges
+
+  bool can_reduce_this_phi(PhiNode* ophi) const;
```
On naming: IMO referring to "this" doesn't help, but adds noise. If you drop it ("can_reduce_this_phi" => "can_reduce_phi"), it's still clear what the method does.

iwanowww commented on src/hotspot/share/code/debugInfo.cpp:
```diff
@@ -191,6 +216,104 @@ void ObjectValue::print_fields_on(outputStream* st) const {
+    return _selected;
+  } else {
+    assert(selector < _possible_objects.length(), "sanity");
+    _selected = (ObjectValue*) _possible_objects.at(selector);
```
Any particular reason to reuse `ObjectValue` from `_possible_objects` instead of allocating a fresh one (as you do on `selector == -1` bracnh)? I'd prefer `ObjectMergeValue::select()` to always allocate a fresh `ObjectValue` when converting `ObjectMergeValue` + `ObjectMergeCandidateValue` into `ObjectValue`.

JohnTortugo replied:
@iwanowww - may I ask why always allocating a fresh object might be better than returning a pointer to a previous "selected" object?

iwanowww replied:
I don't mind there's caching happening if it gives any noticeable benefit. As of now, the code around doesn't care, probably, because it is allocated in resource arena.

What I'm against is repurposing existing instances: don't modify a candidate object into a "real object", allocate a fresh one instead.

JohnTortugo replied:
Thanks for clarifying. There is one scenario where turning the candidate into a "real object" simplify the implementation _greatly_. The scenario is when the ObjectValue is not just a candidate. I.e., the ObjectValue is also used independently of the merge. Example:

```
Point p = new Point();
Point q = new Point();
if (cond) p = q;
trap(p, q);
```

Second issue, is that allocating a fresh ObjectValue will require copying the array of field values from the candidate object to the newly allocated object. That's not a big issue, just pointing that out, though.

I propose that we allocate a fresh ObjectValue if the candidate is just a candidate (not used independent of merge) and if the candidate is not just a candidate we return the existing ObjectValue (turned 'real object'). I have that implemented, I can push it for you to take a look. What do you think?

iwanowww replied:
> ObjectValue is not just a candidate. I.e., the ObjectValue is also used independently of the merge.

And now I'm wondering how it all plays with `is_only_merge_candidate()()`/`set_merge_candidate()`... 

Is it possible for an `ObjectValue` to be shared between multiple merges?

JohnTortugo replied:
> Is it possible for an ObjectValue to be shared between multiple merges?

When I posted my previous comment I thought that could happen. But now I realize that in the current implementation that won't happen: an ObjectValue is created for a combination of Phi x SafePointNode.

However, one situation would _require_ sharing the ObjectValue in multiple merges: when different merges share at least one SR input are used as debug info _in the same_ SafePointNode. It's required because in the same SafePointNode all ObjectValues coming from same Allocate needs to have the same value. I think the example below will trigger that - I'll check and patch the current implementation to not RAM in that case.

```
Point p = new Point();
Point q = new Point();
Point r = new Point();
if (cond_one) p = q;
if (cond_two) r = q;
trap(p, q, r);
```

iwanowww reviewed (commented):
Nice work, Cesar! I like how the patch shapes now.

I'm not done with the review yet, but decided to share the comments I have so far.

iwanowww commented on src/hotspot/share/code/debugInfo.hpp:
```diff
@@ -165,13 +187,59 @@ class ObjectValue: public ScopeValue {
   bool                       is_cached() const          { return _cached; }
   void                       set_cached(bool cached)    { _cached = cached; }
-  AutoBoxObjectValue(int id, ScopeValue* klass) : ObjectValue(id, klass), _cached(false) { }
+  AutoBoxObjectValue(int id, ScopeValue* klass, bool only_merge_candidate = false) : ObjectValue(id, klass, only_merge_candidate), _cached(false) { }
```
Any particular reason to allow `AutoBoxObjectValue` to be a merge candidate?

JohnTortugo commented on src/java.base/share/classes/java/security/AccessController.java:
```diff
@@ -778,6 +778,13 @@ private static boolean isPrivileged() {
+        // situations, after inlining, 'result' (or part of a possibly
+        // allocation merge Phi leading to it) might become NonEscaping and get
+        // scalar replaced. The call below enforces 'result' to always escape.
+        ensureMaterializedForStackWalk(result);
```
Totally missed that!

JohnTortugo commented:
I have an update to this PR to make it possible to scalar replace allocations when the Phi is used in a CmpP (not for all cases). Is there any objection to me pushing these changes? I.e., will it complicate any ongoing review?

JohnTortugo commented:
@vnkozlov - Please let me know if you have further questions.

iwanowww commented:
It took longer than I expected, but I finished looking into debug info. 

A couple of minor comments first:
  * Please, ensure that the AllocationMergesTests.java has cases to trigger the case when SRs and NSRs meet at a merge point. I was not able to provoke it with the unit test.
  * diagnostic output becomes much harder to read (sample output follows).

Sample output:

- ordniary SR case
```
   Expression stack
    - @0: obj: ID=1335, only_merge_candidate=0, skip_field_assignment=0, N.Fields=4, klass: java.lang.String
Fields: 0, 0, 0, nullptr
...
   Objects
obj: ID=1335, only_merge_candidate=0, skip_field_assignment=0, N.Fields=4, klass: java.lang.String 
Fields: 0, 0, 0, nullptr 
```

- mixed merge case:
```
ScopeDesc(pc=0x00000001080bc664 offset=1824):
   java.lang.String::substring@8 (line 2830)
   Locals 
    - l0: merge: ID=1781, N.Candidates=1
...
   Objects
merge: ID=1781, N.Candidates=1obj: ID=1782, only_merge_candidate=1, skip_field_assignment=0, N.Fields=4, klass: java.lang.String 
Fields: 0, 0, 0, reg rfp [58],oop
```

iwanowww commented:
Speaking of debug info design, it seems there's a need for an additional transformation step now.

Originally, all the operations were performed right on the deserialized debug info representation. It was well-justified at first, but slowly accrued with special cases (nulls, autobox, vectors) and merges push it over the limit IMO.

I propose to introduce an additional pass which takes original debug info and, based on current JVM state (`frame` + `RegisterMap`), transforms it into a list of objects to be materialized and a graph of `ScopeValue`s which depend on them. It would isolate preprocessing logic you have scattered across multiple places, simplify rematerialization, make it easier to find out what happens during deoptimizaiton in each particular case. Moreover, it'll enable support for more complex scenarios (e.g., nested merges) which I expect to eventually emerge in followup enhancements.

JohnTortugo commented:
Thank you @iwanowww for taking the time to review this! Please let me ask you some clarifying questions.

> A couple of minor comments first [...]

I'll address those asap! Thanks.

> I propose to introduce an additional pass which takes original debug info [...]

What kind of pass are you referring to exactly? When would this pass run? By "original debug info" you mean the debug information stream?

> It would isolate preprocessing logic you have scattered across multiple places [...]

Which preprocessing logic are you referring to exactly?

iwanowww commented:
The new pass over deserialized debug info would adapt `ScopeDesc::objects()` (initialized by `decode_object_values(obj_decode_offset)` and accesses through `chunk->at(0)->scope()->objects()`) and produce 2 lists:
  * new list of objects which enumerates all scalarized instances which needs to be rematerialized;
  * complete set of objects referenced in the current scope (the purpose `chunk->at(0)->scope()->objects()` serves now).

It should be performed before `rematerialize_objects`.

By preprocessing I mean all the conditional checks before it is attempted to reallocate an `ObjectValue`. By the end of the new pass, it should be enough to just iterate over the new list of scalarized instances in `Deoptimization::realloc_objects`. And after `Deoptimization::realloc_objects` and `Deoptimization::reassign_fields` are over, debug info should be ready to go.

JohnTortugo commented:
@iwanowww - I pushed some changes to address your feedback about the rematerialization part. I added only two more tests for now, but I'm working on adding others.

iwanowww commented:
Very nice, Cesar. I like how the code shapes now.

I verified that the new test cases do trigger SR+NSR scenario. 

How do you test that deoptimization works as expected? 

Diagnostic output is still hard to read. On one hand, it's too verbose when it comes to PcDesc/ScopeDesc sections  ("pc-bytecode offsets" and "scopes")  in nmethod output (enabled either w/ `-XX:+PrintAssembly` or `-XX:CompileCommand=print,...`). On the other hand, it lacks some important details, like `selector` and `merge_ptr` location information which is essential to make sense of debug information at a safepoint in the code.

FTR `_skip_rematerialization` flag is unused now.

Speaking of `_only_merge_candidate` flag, I find it easier about the code when the property being tracked is whether the `ObjectValue` is referenced from corresponding JVM state or not. (Maybe call it `is_root()`?) So, `ScopeDesc::objects_to_rematerialize()` would skip everything not referenced from JVM state, but then unconditionally accept anything returned by `ObjectMergeValue::select()` which doesn't need to adjust the flag before returning selected object. Also, it's safer to track the flag status for every `ObjectValues`, even for `ObjectMergeValue`. 

Are you sure there's no way to end up with nested `ObjectMergeValue`s in presence of iterative EA?

JohnTortugo commented:
> Speaking of _only_merge_candidate flag, I find it easier about the code when the property being tracked is whether the ObjectValue is referenced from corresponding JVM state or not. (Maybe call it is_root()?) So, ScopeDesc::objects_to_rematerialize() would skip everything not referenced from JVM state [...]

@iwanowww - I want to make sure I understood "is_root(sv)" correctly. Are you suggesting to implement it as `ScopeDesc::is_root(ScopeValue* sv)` and the body of the method would just check if the `sv` is referenced in locals/expressions/monitor? Did I get it right?

iwanowww commented:
>  Are you suggesting to implement it as ScopeDesc::is_root(ScopeValue* sv) and the body of the method would just check if the sv is referenced in locals/expressions/monitor? Did I get it right?

I didn't propose exactly that, but I like your idea. I'm not against having it cached on `ScopeValue` side (and serialized in debug info), but implementing it as a query on `ScopeDesc` does look like a better alternative. (If it turns out to matter from performance POV, the check can be then turned into an assert and the cached value is used.)

Maybe call it `ScopeDesc::has_reference_to(ScopeValue* sv)` then.

iwanowww commented:
>> Are you sure there's no way to end up with nested ObjectMergeValues in presence of iterative EA?
> I don't think so. 

Ok. Please, add asserts to catch that situation and a check which bails out compilation (triggering recompilation w/ ReduceAllocationMerges turned off) if it happens with product binaries.

iwanowww commented:
> So, the way I test that deoptimization worked is basically just making sure the tests that "deoptimize" have the same output with RAM enabled and disabled.

Please, enhance `AllocationMergesTests` to cover deoptimizations (e.g., using WhiteBox API or additional run w/ -XX:+DeoptimizeALot) and ensure that tests are sensitive enough to fail when wrong state is rematerialized.

JohnTortugo commented:
Hi @iwanowww - I pushed some changes to address your latest feedback.

> Please, enhance AllocationMergesTests to cover deoptimization (e.g., using WhiteBox API or additional run w/ -XX:+DeoptimizeALot) and ensure that tests are sensitive enough to fail when wrong state is rematerialized.

I added the "+DeoptimizeALot" flag on the tests execution. I also refactored the tests so that each test is executed in the Interpreter and in C2 with the same parameters so that we can confirm that result of the test with RAM enabled is correct.

> Please, add asserts to catch such situation and a check which bails out compilation (triggering recompilation w/ ReduceAllocationMerges turned off) if it happens with product binaries.

I added a new static method `ConnectionGraph::verify_ram_nodes` that does some verification in the inputs and users of RAM nodes. I decided to call the method after each iteration of EA->IGVN->MacroNodeElimination so that we also check that IGVN or `eliminate_macro_nodes` transformations didn't mess with RAM nodes.

> I didn't propose exactly that, but I like your idea. I'm not against having it cached on ScopeValue side (and serialized in debug info), but implementing it as a query on ScopeDesc does look like a better alternative. [...]

I ended up implementing this a little bit different from what I mentioned earlier. I had some problems with the approach that I described before... In current approach I set the `is_root` flag of ObjectValue's right before the object pool is serialized in output.cpp.

iwanowww commented on src/hotspot/share/code/debugInfo.cpp:
```diff
@@ -191,6 +216,104 @@ void ObjectValue::print_fields_on(outputStream* st) const {
+
+void ObjectMergeValue::print_detailed(outputStream* st) const {
+  st->print("merge: ID=%d", _id);
+#ifndef PRODUCT
```
Can you post a sample of the output, please?

Why is it limited to non-product builds? It's valuable irrespective of build flavor.

As I see in `ObjectValue::print_on` and `ScopeDesc::print_on`, you mix `print_on` with `print_fields_on`. Any particular reason for that? You could add `is_object_merge` case in ObjectValue::print_on` instead and extend `ObjectValue::print_fields_on` to cover `ObjectMergeValue` case. I find it hard to reason about `ObjectValue::print_on` vs `ObjectMergeValue::print_on` since it's a non-virtual method. 

 

Also, formatting is broken.

JohnTortugo replied:
I added a few samples below and there are a few more here: https://gist.github.com/JohnTortugo/913523947e08157def6cfebafa7d5daa

Sample 1:

```
Compiled method (c2)     415   24             TestTrapAfterMerge::test (57 bytes)
 total in heap  [0x00007f7b4d03da90,0x00007f7b4d03de18] = 904
 relocation     [0x00007f7b4d03dc00,0x00007f7b4d03dc18] = 24
 main code      [0x00007f7b4d03dc20,0x00007f7b4d03dcb8] = 152
 stub code      [0x00007f7b4d03dcb8,0x00007f7b4d03dcd0] = 24
 oops           [0x00007f7b4d03dcd0,0x00007f7b4d03dce0] = 16
 metadata       [0x00007f7b4d03dce0,0x00007f7b4d03dce8] = 8
 scopes data    [0x00007f7b4d03dce8,0x00007f7b4d03dd50] = 104
 scopes pcs     [0x00007f7b4d03dd50,0x00007f7b4d03de10] = 192
 dependencies   [0x00007f7b4d03de10,0x00007f7b4d03de18] = 8
scopes:
ScopeDesc(pc=0x00007f7b4d03dc3a offset=1a):
   TestTrapAfterMerge::test@-1 (line 3)
ScopeDesc(pc=0x00007f7b4d03dc41 offset=21):
   TestTrapAfterMerge::test@11 (line 5)
ScopeDesc(pc=0x00007f7b4d03dc44 offset=24):
   TestTrapAfterMerge::test@51 (line 12)
ScopeDesc(pc=0x00007f7b4d03dc4a offset=2a):
   TestTrapAfterMerge::test@46 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc52 offset=32):
   TestTrapAfterMerge::test@37 (line 9)
ScopeDesc(pc=0x00007f7b4d03dc57 offset=37):
   TestTrapAfterMerge::test@43 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc61 offset=41):
   TestTrapAfterMerge::test@46 (line 8)  reexecute=true
   Locals
    - l0: empty
    - l1: empty
    - l2: reg rbx [6],int
    - l3: empty
    - l4: merge: ID=26
    - l5: reg r11 [22],int
   Objects
    - 0: merge: ID=26, selector="reg r10 [20],int", merge_pointer="nullptr", candidate objs=[27, 28]
    - 1: obj: ID=27, is_root=0, N.Fields=1, klass: Point 
Fields: reg r8 [16],int
    - 2: obj: ID=28, is_root=0, N.Fields=1, klass: Point 
Fields: reg rcx [2],int
ScopeDesc(pc=0x00007f7b4d03dc63 offset=43):
   TestTrapAfterMerge::test@46 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc6c offset=4c):
   TestTrapAfterMerge::test@34 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc71 offset=51):
   TestTrapAfterMerge::test@55 (line 12)
```

- Sample2:

```
Compiled method (c2)     443   24             TestManys::test (41 bytes)
 total in heap  [0x00007f35e9155b90,0x00007f35e9155e78] = 744
 relocation     [0x00007f35e9155d00,0x00007f35e9155d18] = 24
 main code      [0x00007f35e9155d20,0x00007f35e9155d88] = 104
 stub code      [0x00007f35e9155d88,0x00007f35e9155da0] = 24
 oops           [0x00007f35e9155da0,0x00007f35e9155db0] = 16
 metadata       [0x00007f35e9155db0,0x00007f35e9155db8] = 8
 scopes data    [0x00007f35e9155db8,0x00007f35e9155e10] = 88
 scopes pcs     [0x00007f35e9155e10,0x00007f35e9155e70] = 96
 dependencies   [0x00007f35e9155e70,0x00007f35e9155e78] = 8
scopes:
ScopeDesc(pc=0x00007f35e9155d3a offset=1a):
   TestManys::test@-1 (line 57)
ScopeDesc(pc=0x00007f35e9155d42 offset=22):
   TestManys::test@11 (line 59)
ScopeDesc(pc=0x00007f35e9155d58 offset=38):
   TestManys::test@25 (line 63)
   Locals
    - l0: empty
    - l1: empty
    - l2: empty
    - l3: empty
    - l4: empty
    - l5: empty
    - l6: empty
    - l7: empty
    - l8: merge: ID=26
   Objects
    - 0: merge: ID=26, selector="reg rbp [10],int", merge_pointer="nullptr", candidate objs=[27, 28]
    - 1: obj: ID=27, is_root=0, N.Fields=4, klass: Point 
Fields: stack[36], stack[36], 0, 0
    - 2: obj: ID=28, is_root=0, N.Fields=4, klass: Point 
Fields: 2023, 0, 0, 0
ScopeDesc(pc=0x00007f35e9155d74 offset=54):
   TestManys::test@25 (line 63)
```

- Sample3:
```
Compiled method (c2)     436   24             TestMultiSFO::test (48 bytes)
 total in heap  [0x00007f1df5155590,0x00007f1df5155850] = 704
 relocation     [0x00007f1df5155700,0x00007f1df5155718] = 24
 main code      [0x00007f1df5155720,0x00007f1df5155788] = 104
 stub code      [0x00007f1df5155788,0x00007f1df51557a0] = 24
 oops           [0x00007f1df51557a0,0x00007f1df51557b0] = 16
 metadata       [0x00007f1df51557b0,0x00007f1df51557b8] = 8
 scopes data    [0x00007f1df51557b8,0x00007f1df51557f8] = 64
 scopes pcs     [0x00007f1df51557f8,0x00007f1df5155848] = 80
 dependencies   [0x00007f1df5155848,0x00007f1df5155850] = 8
scopes:
ScopeDesc(pc=0x00007f1df515573a offset=1a):
   TestMultiSFO::test@-1 (line 12)
ScopeDesc(pc=0x00007f1df515575c offset=3c):
   TestMultiSFO::test@28 (line 19)
   Locals
    - l0: empty
    - l1: empty
    - l2: empty
    - l3: merge: ID=14
    - l4: obj: ID=15, is_root=1, N.Fields=2, klass: TestMultiSFO$Point 
Fields: stack[12], stack[8]

   Objects
    - 0: merge: ID=14, selector="reg rbp [10],int", merge_pointer="nullptr", candidate objs=[15, 16]
    - 1: obj: ID=15, is_root=1, N.Fields=2, klass: TestMultiSFO$Point 
Fields: stack[12], stack[8]
    - 2: obj: ID=16, is_root=0, N.Fields=2, klass: TestMultiSFO$Point 
Fields: stack[8], stack[12]
ScopeDesc(pc=0x00007f1df5155778 offset=58):
   TestMultiSFO::test@28 (line 19)
```

JohnTortugo replied:
> Why is it limited to non-product builds? It's valuable irrespective of build flavor.

This is because `print_on` in `AnyObj` is only defined in non-product builds. I based implementation of `ObjectMergeValue::print_on` on `ObjectValue::print_on`. In `ObjectValue::print_on` fields aren't printed in product builds.

> Any particular reason for that? You could add is_object_merge case in ObjectValue::print_oninstead and extendObjectValue::print_fields_onto coverObjectMergeValue case.

I'll do that then.

> Also, formatting is broken.

Can you please share an example? If you mean the tabs on lines 303/304/306/307 I added those because I thought would make the code easier to read, but if you want I can definitely remove that.

iwanowww replied:
>  If you mean the tabs on lines 303/304/306/307

Yes, it confused me. As an alternative, you could put selector and merge_pointer-related statements on the same line, but I'm not sure how much it improves readability:
```
  st->print(", selector=\""); _selector->print_on(st); st->print("\"");
```

iwanowww replied:
A couple of suggestions about the output:
* `merge`: it's clearer to call it `merge_obj`
* `obj` vs `merge` output: obj output is duplicated in ScopeDesc entries and Objects sections; before it was a short version printed in Locals/Expressions and all the details were included in Objects; I like to see field locations in the short version, but including everything looks way too much IMO;
* it makes sense to include selector and merge_pointer info in short version, but `is_root` can be omitted

JohnTortugo replied:
Thanks @iwanowww . Does the output below look good to you? It prints ObjectValue in the same format as it was before this PR and only print details of the merge in the "Objects" section. Is there other output section that you think needs to be adjusted?

```
Compiled method (c2)     436   24             TestMultiSFO::test (48 bytes)
 total in heap  [0x00007f1df5155590,0x00007f1df5155850] = 704
 relocation     [0x00007f1df5155700,0x00007f1df5155718] = 24
 main code      [0x00007f1df5155720,0x00007f1df5155788] = 104
 stub code      [0x00007f1df5155788,0x00007f1df51557a0] = 24
 oops           [0x00007f1df51557a0,0x00007f1df51557b0] = 16
 metadata       [0x00007f1df51557b0,0x00007f1df51557b8] = 8
 scopes data    [0x00007f1df51557b8,0x00007f1df51557f8] = 64
 scopes pcs     [0x00007f1df51557f8,0x00007f1df5155848] = 80
 dependencies   [0x00007f1df5155848,0x00007f1df5155850] = 8
scopes:
ScopeDesc(pc=0x00007f1df515573a offset=1a):
   TestMultiSFO::test@-1 (line 12)
ScopeDesc(pc=0x00007f1df515575c offset=3c):
   TestMultiSFO::test@28 (line 19)
   Locals
    - l0: empty
    - l1: empty
    - l2: empty
    - l3: merge_obj[14]
    - l4: obj[15]

   Objects
    - merge_obj[14], selector="reg rbp [10],int", merge_pointer="nullptr", candidate_objs=[15, 16]
    - obj[15], is_root=1, klass: TestMultiSFO$Point 
	  Fields: stack[12], stack[8]
    - obj[16], is_root=0, klass: TestMultiSFO$Point 
	  Fields: stack[8], stack[12]
```

iwanowww replied:
Thanks, it looks much better now (except the position in Objects array is missing). 

It makes sense to mention `is_root` for merge_obj case even though it's always equals to '1`. Also, make merge_pointer optional and omit it when its value is null.

BTW instead of printing `is_root=0/1`, you can introduce a more compact notation for and mark relevant lines with a single symbol:
```
    - 0: R merge_obj[14], selector="reg rbp [10],int", merge_pointer="nullptr", candidate_objs=[15, 16]
    - 1: R obj[15], is_root=1, klass: TestMultiSFO$Point 
 	        Fields: stack[12], stack[8]
    - 2:   obj[16], is_root=0, klass: TestMultiSFO$Point 
	        Fields: stack[8], stack[12]
```

JohnTortugo replied:
Sounds good. I'll make the changes and push them asap.

iwanowww commented on src/hotspot/share/opto/output.cpp:
```diff
@@ -1061,6 +1086,32 @@ void PhaseOutput::Process_OopMap_Node(MachNode *mach, int current_offset) {
+          bool is_root = locarray->contains(ov) || exparray->contains(ov);
+
+          if (!is_root) {
+            for (int k = 0; k < monarray->length(); k++) {
```
I suggest to turn the lookup over `monarray` into a helper method and call it along with `locarray` and `exparray` checks:
```
bool is_root = locarray->contains(ov) || exparray->contains(ov) || contains_as_owner(monarray, ov);
```

JohnTortugo commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2321,9 +2327,11 @@ void Compile::Optimize() {
-        if (failing())  return;
       }
+
+      NOT_PRODUCT(ConnectionGraph::verify_ram_nodes(this, root());)
```
Duh. I'll fix that.

JohnTortugo commented on src/hotspot/share/code/debugInfo.cpp:
```diff
@@ -191,6 +216,104 @@ void ObjectValue::print_fields_on(outputStream* st) const {
+
+    // Set it to true so that the object will get rematerialized
+    if (!_selected->is_root()) {
+      _selected->set_root(true);
```
I think you're right, there is no need for that. I'll remove/refactor that and run tests again.

JohnTortugo commented:
@iwanowww - I pushed some changes to address your feedback. Please let me know if you have any more comments.

JohnTortugo commented:
@iwanowww - I want to clarify expectations with my colleagues so I have to ask you how much left are there for you to review and whether there is some part of this PR that you're worried about in terms of correctness/performance/etc?

iwanowww commented:
Overall, I like how this patch shapes. 

I need to go through share/opto changes (so far, I did only a shallow review of that part), but the rest looks good. I plan to submit functional and performance testing over the weekend.

JohnTortugo commented:
Thank you for letting me know!

iwanowww commented on test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/AllocationMergesTests.java:
```diff
@@ -0,0 +1,1412 @@
+
+    public static void main(String[] args) {
+        TestFramework.runWithFlags("-XX:+ReduceAllocationMerges",
+                                   "-XX:+TraceReduceAllocationMerges",
```
`TraceReduceAllocationMerges` and `DeoptimizeALot` are not available in product binaries, so the test fails there.
You need to either limit the test to debug builds only or add `-XX:+IgnoreUnrecognizedVMOptions`. 

If `ReduceAllocationMerges` is turned into diagnostic flag, you need to specify `-XX:+UnlockDiagnosticVMOptions`.

iwanowww commented:
Overall, the testing went well. (It discovered some minor issues I commented about.)
I'm rerunning some benchmarks which reported suspicious results. Will keep you posted.

iwanowww commented:
Testing results (both functional and performance) are good.

In addition, I tested with a `C.failure_reason_is(retry_no_reduce_allocation_merges()) == true` guarantee and there were no failures observed.

Once you address my latest comments I'll mark the PR as reviewed.

JohnTortugo commented:
Thank you once more for the comments @iwanowww . I’ll address them asap. 

Can I ask what requirements are there for a product flag?

iwanowww commented:
> Can I ask what requirements are there for a product flag?

Product flags are treated as part of public API of the JVM. So, changes in behavior have to through CSR process. Also, a product flag has to be deprecated/obsoleted first before it can be removed which takes multiple releases to happen. Better to avoid introducing new product flags unless it is well-justified or necessary.

iwanowww commented on src/hotspot/share/opto/c2_globals.hpp:
```diff
@@ -470,6 +470,12 @@
   develop(bool, TracePostallocExpand, false, "Trace expanding nodes after"  \
           " register allocation.")                                          \
                                                                             \
+  product(bool, ReduceAllocationMerges, true,                               \
```
You can look at it in the following way: since the flag is set to true by default, the feature is unconditionally available in product binaries. The only reason to explicitly specify the flag is to turn the optimization off and it may be needed to diagnose VM crashes or performance regressions. 

So, maybe C2 should check a compiler directive (and not a global flag) to be able to control the optimization up to per-method granularity.

JohnTortugo commented on src/hotspot/share/opto/c2compiler.cpp:
```diff
@@ -141,6 +145,12 @@ void C2Compiler::compile_method(ciEnv* env, ciMethod* target, int entry_bci, boo
       }
+      if (C.failure_reason_is(retry_no_reduce_allocation_merges())) {
+        assert(do_reduce_allocation_merges, "must make progress");
+        do_reduce_allocation_merges = false;
```
I created a new [work item](https://bugs.openjdk.org/browse/JDK-8310980) to track this work.

iwanowww reviewed (approved):
The patch looks good.

I resubmitted testing with the latest version and the results are clean.

vnkozlov reviewed (approved):
The final version looks good to me too.

JohnTortugo commented:
@iwanowww @vnkozlov - can I ask one of you to sponsor this PR?

vnkozlov commented:
/sponsor

ijuma commented:
If I understand correctly, this change was backported to older Microsoft OpenJDK versions. Is there a plan to do the same for upstream?
